aten/src/ATen/native/AdaptiveMaxPooling2d.cpp

```diff
@@ -0,0 +1,362 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+#include <tuple>
+
+
+namespace at {
+namespace native {
+
+namespace {
+
+inline int start_index(int a, int b, int c) {
+  return (int)std::floor((float)(a * c) / b);
+}
+
+inline int end_index(int a, int b, int c) {
+  return (int)std::ceil((float)((a + 1) * c) / b);
+}
+
+// #define START_IND(a,b,c) a * c / b
+// #define END_IND(a,b,c)  (a + 1) * c / b + ((a + 1) * c % b > 0)?1:0
+
+// 4d tensor B x D x H x W
+
+template <typename scalar_t>
+static void adaptive_max_pool2d_out_frame(
+        scalar_t *input_p,
+        scalar_t *output_p,
+        int64_t *ind_p,
+        int64_t sizeD,
+        int64_t isizeH,
+        int64_t isizeW,
+        int64_t osizeH,
+        int64_t osizeW,
+        int64_t istrideD,
+        int64_t istrideH,
+        int64_t istrideW)
+{
+  int64_t d;
+#pragma omp parallel for private(d)
+  for (d = 0; d < sizeD; d++)
+  {
+    /* loop over output */
+    int64_t oh, ow;
+    for(oh = 0; oh < osizeH; oh++)
+    {
+      int istartH = start_index(oh, osizeH, isizeH);
+      int iendH   = end_index(oh, osizeH, isizeH);
+      int kH = iendH - istartH;
+
+      for(ow = 0; ow < osizeW; ow++)
+      {
+        int istartW = start_index(ow, osizeW, isizeW);
+        int iendW   = end_index(ow, osizeW, isizeW);
+        int kW = iendW - istartW;
+
+        /* local pointers */
+        scalar_t *ip = input_p   + d*istrideD + istartH*istrideH + istartW*istrideW;
+        scalar_t *op = output_p  + d*osizeH*osizeW + oh*osizeW + ow;
+        int64_t *indp = ind_p   + d*osizeH*osizeW + oh*osizeW + ow;
+
+        /* compute local max: */
+        int64_t maxindex = -1;
+        scalar_t maxval = -std::numeric_limits<float>::max();
+        int ih, iw;
+        for(ih = 0; ih < kH; ih++)
+        {
+          for(iw = 0; iw < kW; iw++)
+          {
+            scalar_t val = *(ip + ih*istrideH + iw*istrideW);
+            if ((val > maxval) || std::isnan(val))
+            {
+              maxval = val;
+              maxindex = (ih+istartH)*isizeW + (iw+istartW);
+            }
+          }
+        }
+
+        /* set output to local max */
+        *op = maxval;
+
+        /* store location of max */
+        *indp = maxindex;
+      }
+    }
+  }
+}
+
+void adaptive_max_pool2d_out_cpu_template(
+          Tensor& output,
+          Tensor& indices,
+          const Tensor& input,
+          IntArrayRef output_size)
+{
+  int dimW = 2;
```

Comment:
Well, `dimW` only takes the values 2 or 3, so `int` is fine.

Reply:
I suppose there is a very nitpicky argument to be made about how it might be better to use `int64_t` because it prevents someone else from copypasting your code and then committing a "oops I used 32-bit indexing when I should have used 64-bit indexing error". I do think this is pretty benign, so I'll leave it up to you @skrah.